knative-operator/pkg/controller/knativeserving/consoleclidownload/consoleclidownload.go

```diff
@@ -1,51 +1,310 @@
+}
+
+// manifestPath returns kn ConsoleCLIDownload deployment resource manifest path
 func manifestPath() string {
-	knConsoleCLIDownload := os.Getenv("CONSOLE_DOWNLOAD_MANIFEST_PATH")
-	if knConsoleCLIDownload == "" {
-		return defaultKnConsoleCLIDownload
+	return os.Getenv("CONSOLECLIDOWNLOAD_MANIFEST_PATH")
+}
+
+func replaceKnCLIArtifactsImage(image string, scheme *runtime.Scheme) mf.Transformer {
+	return func(u *unstructured.Unstructured) error {
+		if u.GetKind() == "Deployment" {
+			deploy := &appsv1.Deployment{}
+			if err := scheme.Convert(u, deploy, nil); err != nil {
+				return fmt.Errorf("failed to convert unstructured obj to Deployment: %w", err)
+			}
+
+			containers := deploy.Spec.Template.Spec.Containers
+			for i, container := range containers {
+				if container.Name == knDownloadServer && container.Image != image {
+					log.Info("Replacing", "deployment", container.Name, "image", image)
+					containers[i].Image = image
+					break
+				}
+			}
+
+			if err := scheme.Convert(deploy, u, nil); err != nil {
+				return fmt.Errorf("failed to convert Deployment obj to unstructured: %w", err)
+			}
+		}
+		return nil
+	}
+}
+
+func setOwnerAnnotations(instance *servingv1alpha1.KnativeServing) mf.Transformer {
+	return func(u *unstructured.Unstructured) error {
+		u.SetAnnotations(map[string]string{
+			OwnerName:      instance.Name,
+			OwnerNamespace: instance.Namespace,
+		})
+		return nil
+	}
+}
+
+// populateKnConsoleCLIDownload populates kn ConsoleCLIDownload object and its SPEC
+// using route's baseURL
+func populateKnConsoleCLIDownload(baseURL string, instance *servingv1alpha1.KnativeServing) *consolev1.ConsoleCLIDownload {
+	anno := make(map[string]string)
+	if instance != nil {
+		anno = map[string]string{
+			OwnerName:      instance.Name,
+			OwnerNamespace: instance.Namespace,
+		}
+	}
+	return &consolev1.ConsoleCLIDownload{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:        knCLIDownload,
+			Annotations: anno,
+		},
+		Spec: consolev1.ConsoleCLIDownloadSpec{
+			DisplayName: "kn - OpenShift Serverless Command Line Interface (CLI)",
+			Description: "The OpenShift Serverless client `kn` is a CLI tool that allows you to fully manage OpenShift Serverless Serving and Eventing resources without writing a single line of YAML.",
+			Links: []consolev1.Link{
+				consolev1.Link{
+					Text: "Download kn for Linux",
+					Href: baseURL + "/amd64/linux/kn-linux-amd64.tar.gz",
+				},
+				consolev1.Link{
+					Text: "Download kn for macOS",
+					Href: baseURL + "/amd64/macos/kn-macos-amd64.tar.gz",
+				},
+				consolev1.Link{
+					Text: "Download kn for Windows",
+					Href: baseURL + "/amd64/windows/kn-windows-amd64.zip",
+				},
+			},
+		},
+	}
+}
+
+// Function copied from github.com/openshift/console-operator/pkg/console/subresource/route/route.go and modified
+func getCanonicalHost(route *routev1.Route) string {
+	for _, ingress := range route.Status.Ingress {
+		if ingress.RouterName != defaultIngressController {
```

Comment:
Why do we care about this check? Does this mean that this won't work with non-default ingress controllers?

I feel like we should just return the first non-empty `ingress.Host` 🤷‍♂️

Reply:
I am not sure about this, I copied it over to avoid vendoring. I will remove it for now and lets see how it goes ;-)